src/js/extensions/paste.js

```diff
@@ -86,48 +89,128 @@
+
+            isKeyBoardPaste = (new Date().getTime() - keyboardPasteTimeStamp - clipboardDelay) < 1000;
+
+            plainTextMode = !this.cleanPastedHTML || this.forcePlainText;
+
+            window.console.log('handlePaste > plainTextMode', plainTextMode);
+
+            if (event.isDefaultPrevented) {
+                this.removePasteBin();
+                return;
             }
 
-            if (event.clipboardData &&
-                    event.clipboardData.getData &&
-                    !event.defaultPrevented) {
+            // Not a keyboard paste prevent default paste and try to grab the clipboard contents using different APIs
+            if (!isKeyBoardPaste) {
                 event.preventDefault();
+            }
+
+            setTimeout(function () {
+                window.console.log('handlePaste > hasContentType', that.hasContentType(clipboardContent, 'text/html'));
+
+                // Grab HTML from Clipboard API or paste bin as a fallback
+                if (that.hasContentType(clipboardContent, 'text/html')) {
+                    content = clipboardContent['text/html'];
+                } else {
+                    content = that.getPasteBinHtml();
+
+                    // If paste bin is empty try using plain text mode
+                    // since that is better than nothing right
+                    if (content === pasteBinDefaultContent) {
+                        plainTextMode = true;
+                    }
+                }
+
+                content = that.trimHtml(content);
+
+                window.console.log('handlePaste > content', content);
+
+                // WebKit has a nice bug where it clones the paste bin if you paste from for example notepad
+                // so we need to force plain text mode in this case
+                if (this.pasteBinElm && this.pasteBinElm.firstChild && this.pasteBinElm.firstChild.id === 'mcepastebin') {
+                    plainTextMode = true;
+                }
+
+                that.removePasteBin();
+
+                // If we got nothing from clipboard API and pastebin then we could try the last resort: plain/text
+                if (!content.length) {
+                    plainTextMode = true;
+                }
+
+                // Grab plain text from Clipboard API or convert existing HTML to plain text
+                if (plainTextMode) {
+                    // Use plain text contents from Clipboard API unless the HTML contains paragraphs then
+                    // we should convert the HTML to plain text since works better when pasting HTML/Word contents as plain text
+                    if (that.hasContentType(clipboardContent, 'text/plain') && content.indexOf('</p>') === -1) {
+                        content = clipboardContent['text/plain'];
+                    } else {
+                        // What is innerText supposed to do? (It isn't defined)
+                        //content = MediumEditor.util.innerText(content);
```

Comment:
I found nothing about that on my work dev machine. I guess it was to retrieve the html from `content`? I can't remember